Introduce delayed compile via `eager_then_compile` stance

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147983

Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
[experimental] delayed compile

[ghstack-poisoned]

bobrenjc93 committed:
Update on "[experimental] delayed compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "[experimental] delayed compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/decorators.py:
```diff
@@ -114,7 +114,11 @@ def __init__(
+            and force_backend is not None
+            and stance != "default"
+        ):
             raise RuntimeError("non-default stance cannot have force_backend set")
```
Update this error message? (also could cleanup the above condition to `force_backend is not None and stance not in (...)`)

bobrenjc93 replied:
Whoops I meant to remove this code

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -166,6 +184,40 @@ def _callback_from_stance(callback):
+    elif _stance.stance == "eager_then_compile":
+        if callback not in (False, None):
+
+            def foo(*args, **kwargs):
```
Match the convert frame function protocol here for readability

bobrenjc93 replied:
Hmmm but I need args kwargs later on in the function. One unpacking seems nicer and much less verbose than unpacking everything?

williamwen42 replied:
Ok, fair

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -166,6 +184,40 @@ def _callback_from_stance(callback):
+
+            def foo(*args, **kwargs):
+                frame = args[0]
+                key = frame.f_code.co_filename + str(frame.f_code.co_firstlineno)
```
Is this key computation correct? Would resume functions have the same key as the original function?

bobrenjc93 replied:
Can you elaborate on what you mean by resume function? Is this for graph breaks? Bonus points if you can give a concrete example

williamwen42 replied:
Yes, for graph breaks. Resume functions are generated to continue tracing after a graph break. e.g.
```python
import torch


@torch.compile(backend="eager")
def f(x):
    x = torch.sin(x)
    torch._dynamo.graph_break()
    x = torch.cos(x)
    return x


f(torch.randn(3))
```
gives resume function
```
MODIFIED BYTECODE torch_dynamo_resume_in_f_at_7 /data/users/williamwen/pytorch3/playground3.py line 7 
  7           0 RESUME                   0
              2 LOAD_GLOBAL             11 (NULL + __compiled_fn_5)
             12 LOAD_FAST                1 (x)
             14 COPY                     1
             16 STORE_FAST               2 (tmp_0)
             18 CALL                     1
             26 UNPACK_SEQUENCE          1
             30 RETURN_VALUE
```
If we graph break multiple times on the same line, we would get the same filename + line number.

I'm not sure if key collisions between original and resume functions are intentional or not.

bobrenjc93 replied:
I see. Yeah in this case it doesn't matter because the first invocation is eager and the second one will fill up the example_inputs before any graph breaks are hit. That being said, if you have a suggestion to make this more future proof and robust, I'm happy to adjust.

I also added a test to verify this behavior.

williamwen42 reviewed (commented):
some comments

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

laithsakka commented on test/dynamo/test_decorators.py:
```diff
@@ -624,6 +624,42 @@ def forward(self, x):
+
+        with torch.compiler.set_stance("eager_then_compile"):
+            fn(1, torch.randn(1), {0: torch.randn(1)})
+            fn(2, torch.randn(2), {0: torch.randn(2)})
```
nit: I mean you only need 2? max 3.

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 committed:
Update on "Introduce delayed compile via `eager_then_compile` stance"


Recently I've been experimenting with introducing new APIs to delay compile as a way to reduce compile times while improving the ergonomics of using dynamic shapes. The high level idea is to run the first invocation of compile in eager, save the example inputs, and on the second invocation we can derive the dynamism in the inputs so that we don't need to waste our time doing a compile with static shapes (which is the status quo today with automatic dynamic).

Another benefit of this is most users no longer need to annotate their inputs with mark_dynamic and mark_unbaked calls since we can derive the dynamism on the very first call. Additionally we get dynamic ints out of the box in this new regime.

This PR implements this idea through the set_stance APIs. In particular it introduces a new `eager_then_compile` stance.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
